content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,711 @@
+	while let (Some(arg), Some(value)) = (args.next(), args.next()){
+		if arg == "--name" {
+			name = Some(value.to_string());
+		}
+		if arg == "--answer" {
+			answer = Some(value.parse::<i32>.unwrap()),
+		}
+	};
+
+	Args {
+		name: name.expect("'--name' not given"),
+		answer: answer.expect("'--answer' not given"),
+	}
+}
+```
+
+What if you could use functional programming to just describe what you want:
+
+```rust
+fn parse() -> Args {
+    let name = Arg("--name").parse(|n| n.to_string())
+    let answer = Arg("--answer").parse(|a| a.parse::<i32>)
+    let Args = name
+      .zip(answer)
+      .map(|name, answer| Args { name, answer })
+      .run()
+      .unwrap()
+}
+```
+
+Now lets see how we get there.
+
+# Parsing command line options
+
+The main goal for this tutorial is to implement a basic library to parse command line options.
+If you used `cargo` you are probably familiar with expressions like this:
+
+```console
+$ cargo build --bin hello      # build a binary "hello"
+$ cargo check --package server # check if package "server" contains any issues
+```
+
+Parsing command line options in general is a wide problem and to make it more manageable this
+tutorial is going to look only at option arguments with a long name: `--bin hello` and
+`--package server` parts. This approach however can parse pretty much anything,
+[`bpaf`](https://crates.io/crates/bpaf) is a finished library that uses it.
+
+## Humble beginning
+
+Tutorial starts with some magical handwavey way of taking things from `std::env::args()` and placing
+them into variables and introduces a simple way of taking them out of the variables.
+
+```rust
+// a magical container with a value already inside
+struct Magic(&'static str);
+impl Magic {
+    // a way to get a value out of the container
+    fn run(self) -> String {
+        self.0.into()
+    }
+}
+
+#[test]
+fn sample() {
+    // --name Bob
+    let name = Magic("Bob");
+    let _result = name.run(); // "Bob"
+}
+```
+
+At this point parser can consume multiple string arguments using `Magic` and `Magic::run`.
+
+## Adding more types
+
+Next step would be to allow to have arguments of different types. Consider filenames. While
+most of the functions would take a regular string reference as a filename - not every `String` is
+a valid filename and not every filename is a valid `String`. It's a good practice to keep
+`String` for strings and `PathBuf` for filenames. Type can also serve as a documentation for
+users.
+
+To be able to consume `PathBuf` from user's input parser needs a way to represent it in the first
+place. One approach would be to make `Magic` a generic datatype:
+
+```rust
+struct Magic<T>(T);
+impl<T> Magic<T> {
+    fn run(self) -> T {
+        self.0
+    }
+}
+```
+
+This allows to represent values and get them out but since handwavy magic creates variables of
+type `Magic<&'static str>` parser still needs a way to change the type to `Magic<PathBuf>` or
+any other type.
```

Comment:
```suggestion
Generics allow us to represent and extract values of any type, but since our handwavy magic only creates variables of
type `Magic<&'static str>`, the parser still needs a way to change the type to another such as `Magic<PathBuf>`.
```